src/stream/src/executor/merge.rs

```diff
@@ -200,12 +217,28 @@ impl Executor for MergeExecutor {
     }
 }
 
+#[derive(Default)]
+struct StagedWatermarks {
+    in_heap: bool,
+    staged: VecDeque<Watermark>,
+}
+
+struct UntransferredWatermarks {
+    // We store the smallest watermark of each upstream, because the next watermark to emit is
+    // among them.
+    pub first_untransferred_watermarks: BinaryHeap<Reverse<(Watermark, ActorId)>>,
+    // We buffer other watermarks of each upstream. The next-to-smallest one will become the
+    // smallest and be moved into heap
+    pub other_untransferred_watermarks: HashMap<ActorId, StagedWatermarks>,
+}
+
 pub struct SelectReceivers {
     blocks: Vec<BoxedInput>,
     upstreams: Vec<BoxedInput>,
     barrier: Option<Barrier>,
     last_base: usize,
     actor_id: u32,
+    untransferred_watermarks: BTreeMap<usize, UntransferredWatermarks>,
```

Comment:
Could you please add some doc to show that this map is indexed by the column index?

Reply:
> Could you please add some doc to show that this map is indexed by the column index?

resolved